packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -28,8 +29,60 @@ export class AssociationManager {
     };
   }
 
+  public static constrainTagUpdates(portfolio: IPortfolio, product: IProduct, options: TagUpdateConstraintOptions): void {
+    this.validateConstraintDescription(this.prettyPrintAssociation(portfolio, product), options);
+    const association = this.associateProductWithPortfolio(portfolio, product);
+    const constructId = `ResourceUpdateConstraint${association.associationKey}`;
+
+    if (!portfolio.node.tryFindChild(constructId)) {
+      const constraint = new CfnResourceUpdateConstraint(portfolio as unknown as cdk.Resource, constructId, {
+        acceptLanguage: options.messageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allow === false ? 'NOT_ALLOWED' : 'ALLOWED',
+      });
+
+      // Add dependsOn to force proper order in deployment.
+      constraint.addDependsOn(association.cfnPortfolioProductAssociation);
+    } else {
+      throw new Error(`Cannot have multiple tag update constraints for association ${this.prettyPrintAssociation(portfolio, product)}`);
+    }
+  }
+
+  public static deployWithStackSets(portfolio: IPortfolio, product: IProduct, accounts: string[], regions: string[], adminRole: iam.IRole,
+    executionRole: iam.IRole, allowInstanceControl: boolean, options: CommonConstraintOptions) {
+
+    this.validateConstraintDescription(this.prettyPrintAssociation(portfolio, product), options);
+    const association = this.associateProductWithPortfolio(portfolio, product);
+    // Check if a launch role has already been set.
+    if (portfolio.node.tryFindChild(`LaunchRoleConstraint${association.associationKey}`)) {
```

Comment:
I debated breaking out what we call the constructs into an enum again because we cross reference them in certain instances like here.

That OR what I had originally done was just use `LaunchRoleOrStackSetConstraint` and then I wouldn't have a secondary check.  If this is allowed then we can just update error message to a more generic `cannot have more than one launch role or stackset deployment configured`

Reply:
I don't think you need an enum, but rather 2 private helper functions here, `stackSetConstraintLogicalId(associationKey: string): string` and `launchRoleConstraintLogicalId(associationKey: string): string`.